pkg/ebpf/tracer.go

```diff
@@ -765,38 +767,51 @@ func (t *Tracer) DebugNetworkMaps() (*network.Connections, error) {
 // closed ports will be returned.
 // the map will be one of port_bindings  or udp_port_bindings, and the mapping will be one of tracer#portMapping
 // tracer#udpPortMapping respectively.
-func (t *Tracer) populatePortMapping(mp *ebpf.Map, mapping *network.PortMapping) ([]uint16, error) {
-	var key, emptyKey uint16
+func (t *Tracer) populatePortMapping(mp *ebpf.Map, mapping *network.PortMapping) (closed []portBindingTuple, closedNetNs []uint64, err error) {
+	var key, emptyKey portBindingTuple
 	var state uint8
 
-	closedPortBindings := make([]uint16, 0)
-
+	inos := make(map[int]uint64)
 	entries := mp.IterateFrom(unsafe.Pointer(&emptyKey))
 	for entries.Next(unsafe.Pointer(&key), unsafe.Pointer(&state)) {
-		port := key
+		pid := int(key.pid)
+		ino := uint64(key.net_ns)
+		port := uint16(key.port)
+		if ino == 0 {
+			var ok bool
+			if ino, ok = inos[pid]; !ok {
+				var err error
+				if ino, err = util.GetNetNsInoFromPid(t.config.ProcRoot, pid); err != nil {
+					log.Errorf("could not add port mapping (pid: %d port: %d), could not get net ns for pid %d: %s", pid, port, pid, err)
+					continue
+				}
+
+			}
+		}
 
-		mapping.AddMapping(port)
+		inos[pid] = ino
+
+		log.Tracef("adding port mapping for port=%d net_ns=%d", port, ino)
+		mapping.AddMapping(ino, port)
 
 		if isPortClosed(state) {
-			closedPortBindings = append(closedPortBindings, port)
+			closed = append(closed, key)
+			closedNetNs = append(closedNetNs, ino)
 		}
 	}
 
-	return closedPortBindings, nil
+	return closed, closedNetNs, nil
 }
 
 func (t *Tracer) determineConnectionDirection(conn *network.ConnectionStats) network.ConnectionDirection {
+	pm := t.portMapping
 	if conn.Type == network.UDP {
-		if t.udpPortMapping.IsListening(conn.SPort) {
-			return network.INCOMING
-		}
-		return network.OUTGOING
+		pm = t.udpPortMapping
 	}
-
-	if t.portMapping.IsListening(conn.SPort) {
+	if pm.IsListening(uint64(conn.NetNS), conn.SPort) ||
```

Comment:
Just to verify my understanding -- this check is only necessary when we're checking the initial port mappings because the mappings from /proc/net use post-NAT ports and the ports from eBPF use pre-NAT ports ?

Reply:
I don't think it is necessary if NAT info is present. I can switch the order around.